src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/RazorProjectService.cs

```diff
@@ -19,28 +19,142 @@
-    private readonly ILogger _logger = loggerFactory.GetOrCreateLogger<RazorProjectService>();
+    private readonly IRazorProjectInfoDriver _projectInfoDriver;
+    private readonly IProjectSnapshotManager _projectManager;
+    private readonly RemoteTextLoaderFactory _remoteTextLoaderFactory;
+    private readonly IDocumentVersionCache _documentVersionCache;
+    private readonly ILogger _logger;
+
+    private readonly CancellationTokenSource _disposeTokenSource;
+    private readonly Task _initializeTask;
+
+    public RazorProjectService(
+        IProjectSnapshotManager projectManager,
+        IRazorProjectInfoDriver projectInfoDriver,
+        IDocumentVersionCache documentVersionCache,
+        RemoteTextLoaderFactory remoteTextLoaderFactory,
+        ILoggerFactory loggerFactory)
+    {
+        _projectInfoDriver = projectInfoDriver;
+        _projectManager = projectManager;
+        _remoteTextLoaderFactory = remoteTextLoaderFactory;
+        _documentVersionCache = documentVersionCache;
+        _logger = loggerFactory.GetOrCreateLogger<RazorProjectService>();
+
+        // We kick off initialization immediately to ensure that the IRazorProjectService
+        // is hot and ready to go when requests come in.
+        _disposeTokenSource = new();
+        _initializeTask = InitializeAsync(_disposeTokenSource.Token);
+    }
 
-    public Task AddDocumentToMiscProjectAsync(string filePath, CancellationToken cancellationToken)
+    public void Dispose()
     {
-        return _projectManager.UpdateAsync(
-            updater: AddDocumentToMiscProjectCore,
-            state: filePath,
-            cancellationToken);
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
+    }
+
+    private async Task InitializeAsync(CancellationToken cancellationToken)
+    {
+        _logger.LogTrace($"Initializing {nameof(RazorProjectService)}...");
+
+        await _projectInfoDriver.WaitForInitializationAsync().ConfigureAwait(false);
+
+        // Register ourselves as a listener to the project driver.
+        _projectInfoDriver.AddListener(this);
+
+        // Add all existing projects from the driver.
+        foreach (var projectInfo in _projectInfoDriver.GetLatestProjectInfo())
+        {
+            await AddOrUpdateProjectCoreAsync(
+                projectInfo.ProjectKey,
+                projectInfo.FilePath,
+                projectInfo.Configuration,
+                projectInfo.RootNamespace,
+                projectInfo.DisplayName,
+                projectInfo.ProjectWorkspaceState,
+                projectInfo.Documents,
+                cancellationToken)
+                .ConfigureAwait(false);
+        }
+
+        _logger.LogTrace($"{nameof(RazorProjectService)} initialized.");
+
+    }
+
+    Task IOnInitialized.OnInitializedAsync(ILspServices services, CancellationToken cancellationToken)
+#pragma warning disable VSTHRD003 // Avoid awaiting foreign Tasks
+        => _initializeTask;
```

Comment:
I don't understand what this is doing. This task will complete when `IOnInitialize` happens, but we're returning it from `IOnInitialize*d*`? What is this expected to do?

Reply:
It should only implicitly waits for IOnInitialized if the FileWatcherBasedRazorProjectInfoDriver is used because it needs the workspace root. In the VS case, the driver starts populating immediately from the IProjectSnapshotManager. I'll take another look at IWorkspaceRootPathProvider. However, because it's dependent on getting the workspace root path from initialization, there aren't many options.

Reply:
Ultimately, I think it's fine for `RazorProjectServer` _not_ to implement `IOnInitialized`.